Setup sdk-based embedding iframe route

heypoom opened the pull request:
Closes EMB-320

Creates a new endpoint `/embed/sdk/v1` that serves a minimal page which renders the SDK components based on the embed config. It also exposes a tiny JS library for interacting with the iframe.

You must set the `METASTORE_DEV_SERVER_URL` to `"https://token-check.staging.metabase.com"` and clean the instance, otherwise the token features will not activate.

The e2e tests are moved to #57269 so this PR is a bit easier to review.

### How to verify

You'll need to re-create your instance to make sure the `embedding_iframe_sdk` feature flag is applied.

Setup a HTML file containing this snippet, and serve it (example: `python3 -m http.server 3456`). Don't forget to replace the `instanceUrl` and `apiKey` with yours!

```html
<script src="http://localhost:3000/app/dist/embed.v1.js"></script>
<div id="metabase-embed-container"></div>

<style>
  body {
    margin: 0;
  }

  #metabase-embed-container {
    height: 100vh;
  }
</style>

<script>
  const { MetabaseEmbed } = window["metabase.embed"];

  const embed = new MetabaseEmbed({
    target: "#metabase-embed-container",
    instanceUrl: "http://localhost:3000",
    dashboardId: 1,
    apiKey: "(insert-your-api-key-here)",
    theme: {
      colors: {
        background: "#2d2d3d",
        "text-primary": "#fff",
        "text-secondary": "#b3b3b3",
        "text-tertiary": "#8a8a8a",
        brand: "#ff9900",
      },
    },
  });
</script>
```

Then, follow [the testing plan](https://www.notion.so/metabase/Testing-Plan-New-iframe-embedding-using-SDK-components-1e469354c9018081b08be44571b82453?pvs=4) on Notion. You can also reference the end-to-end tests on #57269

### Demo

![CleanShot 2568-04-30 at 00 37 09@2x](https://github.com/user-attachments/assets/76bb97ef-d59f-49ee-ae29-5612fb94a52b)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR: e2e tests are moved to #57269 to reduce the size of this PR.

heypoom committed:
setup a new embedding_iframe_sdk EE plugin

heypoom committed:
add the interactive embedding provider

heypoom committed:
add the interactive embed route component

heypoom committed:
add the iframe interactive embedding route to routes-embed

heypoom committed:
change the duplicate visualization registration error to a warning

heypoom committed:
add the embed.js script

heypoom committed:
gate the public reducer plugins via an EE plugin

heypoom committed:
revert me: mock the embedding_iframe_sdk token feature for testing

heypoom committed:
check for invalid base64 strings

heypoom committed:
small cleanup for auth

heypoom committed:
use postMessage to receive settings

heypoom committed:
implement postMessage for settings in embed.js

heypoom committed:
sync the iframe embed routes

heypoom committed:
fix sdk iframe not rendering

heypoom committed:
fix embed.js postMessage field

heypoom committed:
scaffold cypress tests

heypoom committed:
use cy.log for main test file

heypoom committed:
cleanup test files

heypoom committed:
add testing flags

heypoom committed:
fix tests not loading iframe

heypoom committed:
remove debug logging

heypoom committed:
remove iframe test page heading

heypoom committed:
update test files

heypoom committed:
remove CLAUDE.md

heypoom committed:
remove e2e tests to move to another branch

heypoom committed:
revert store/auth.ts

heypoom committed:
rename url to instance url

heypoom committed:
merge the settings in embed.updateSettings

heypoom committed:
cleanup jsdoc strings

heypoom committed:
prevent providing both dashboard and question

heypoom committed:
inline the sdk provider

heypoom committed:
tell the redux store we are embedding sdk components

heypoom committed:
update embed route flag

heypoom committed:
remove empty type file

heypoom committed:
leave default-component-theme alone

heypoom committed:
remove PLUGIN_EMBEDDING_IFRAME_SDK.isInteractiveEmbeddingEnabled for now

heypoom committed:
rename embedding route to embed.v2

heypoom committed:
bump version to 1.0.0

heypoom committed:
remove the debug log

heypoom committed:
EmbedSettings can also be a number

heypoom committed:
update route to embed/sdk/v1

heypoom committed:
add hints on why we need to setIsEmbeddingSdk

heypoom committed:
cleanup JSDoc type hints

heypoom committed:
simplify new iframe implementation

heypoom committed:
simplify authConfig condition

heypoom committed:
rename iframeSettings to embedSettings

heypoom committed:
remove metabase.embed.authenticate for now

heypoom committed:
build embed.v1.ts via rspack

heypoom committed:
move the script tag to typescript

heypoom committed:
update copy in enable-embedding-iframe-sdk?

heypoom committed:
pass instanceUrl down

heypoom committed:
remove void return type

heypoom committed:
ensure embed.v1.js is generated on dev

heypoom committed:
update rspack config

heypoom committed:
use iframeEmbedConfig

heypoom committed:
update rspack config to load ts

heypoom committed:
remove old rspack config

heypoom committed:
use swc-loader

heypoom committed:
update rspack library config

heypoom committed:
validate apiKey and instanceUrl

heypoom committed:
update rspack config

heypoom committed:
lift the rspack config to a separate file

heypoom committed:
reformat rspack.sdk-iframe-embed.config.js

heypoom committed:
namespace the rspack config

heypoom committed:
update rspack config

heypoom committed:
fix namespaced rspack config

heypoom committed:
fix hmr issues

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/rspack.embed.config.js:
```diff
@@ -0,0 +1,52 @@
+    libraryTarget: "umd",
+    globalObject: "this",
+  },
+  devServer: { hot: false },
+  watchOptions: {
+    ignored: ["**/node_modules", SCRIPT_TAG_PATH],
+  },
```
HMR is having issues on the script tag where it's generating `hot-update` files at `main` (as we are using `devServer` on the main rspack config) without ever cleaning them up, so we have to ignore the SCRIPT_TAG_PATH. I think it's a rspack bug that happens when we re-export multiple rspack configs and they have differing `devServer` setups

heypoom committed:
update the token feature test

heypoom committed:
update the stats test

heypoom committed:
merge settings in embed.js

heypoom committed:
remove unused isSdkIframeEmbeddingEnabled field

heypoom committed:
add locale support

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute/SdkIframeEmbedRoute.tsx:
```diff
@@ -0,0 +1,95 @@
+}: {
+  settings: SdkIframeEmbedSettings;
+}) => {
+  const { dashboardId, questionId, notebookEditor } = settings;
+
+  return match({ dashboardId, questionId, notebookEditor })
+    .with({ notebookEditor: true }, () => (
```
Please ignore the `notebookEditor` field in the rest of this PR for now. I will delete it and replace it with `template: "exploration"` to align with the product doc (it was still in definition).

The idea is that:
- If you provide `questionId` or `dashboardId`, it uses the default question or dashboard template
- If you provide `template: "exploration"` it shows the notebook editor + results. No `questionId` needed.
- If you provide `template: "xyz"` you can select from a set of predefined templates of questions or dashboards setup. Example: `template: "metabot"`

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/dd26ba4c060680fb173f353e6809ce7230811725

heypoom committed:
rename notebookEditor to template.exploration

heypoom commented on frontend/src/metabase/visualizations/index.ts:
```diff
@@ -60,7 +60,7 @@ export function registerVisualization(visualization: Visualization) {
     }
 
-    throw new Error(
+    console.warn(
```
Visualization seems to be registering more than once. Looking into why this is the case.

heypoom committed:
cleanup rspack config

heypoom committed:
use PLUGIN_EMBEDDING_IFRAME_SDK.isEnabled to enable route

heypoom committed:
Revert "revert me: mock the embedding_iframe_sdk token feature for testing"

This reverts commit 2984cd79c543f5ba7b4038252750460199f32108.

npretto reviewed (dismissed):
I had a few drafted comments but most of them are about the same thing so I'll summarize them here.
From my understanding we're basically using the sdk code (not the bundled sdk) inside a page in the "public/embed" routes.
I think this has a couple of issues:
- we're mixing core app and sdk code, more specifically we're using the sdk code in a different context than then one we usually do which could create issues (I think the reason you needed the custom class for the portal container is because of mixing core app css and the sdk css?)
- the logic for the reducers seems a bit leaky and duplicated, we're mixing embed and sdk reducers instead of just using the store that the sdk already creates automatically, we're also manually setting the flag for the sdk (which is used on a component mount, so some code may already have used that, which scares a bit as it can lead to hard to debug issues).

While this works, I think we can find a cleaner solution that doesn't add this extra "context" to think about. Ideally, we wouldn't need to test *everything* in this context, as this would be just standard sdk + the glue code, and we can just write just enough tests to make sure that the glue code (post message + auth) are working.

I think it'd be more maintainable if we managed to get a real empty page with its own entry point that is just a small wrapper on the sdk

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute/SdkIframeEmbedRoute.module.css:
```diff
@@ -0,0 +1,4 @@
+.SdkIframeEmbedPortalContainer {
```
@npretto Without changing the embed portal container, the portals does not render inside an iframe. Let me capture a screenshot of how it looks like

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,145 @@
+  }
+}
+
+const raiseError = (message: string) => {
```
@npretto `raiseError` throws a new error so it typically shows as an uncaught error in the console of the embedded site, unless it's caught with a `try-catch` by the user. `warn` prints out a warning in the console.

heypoom commented:
@npretto Good points about the use of public stores for the SDK; I tried to use `<MetabaseProvider />` as-is originally in the embed routes but ran into a couple issues. Let me try that again and see if I can get that to work this time.

heypoom committed:
use getSdkStore instead

heypoom committed:
setup a clean entry point

heypoom committed:
fix sdk prefix

heypoom committed:
update rspack config

heypoom committed:
cleanup embed route

heypoom committed:
cleanup embed entrypoint init

heypoom committed:
use /sdk/v1 route instead

heypoom committed:
remove the cached embed-sdk.html

heypoom committed:
fix issues in embed tag

heypoom committed:
rename the event to metabase.embed.setSettings

heypoom committed:
revert sdk plugin fields

heypoom committed:
ignore the hot-update file

heypoom committed:
revert PortalContainer changes

heypoom committed:
revert MetabaseProvider changes

heypoom committed:
unnest SdkIframeEmbedRoute

heypoom committed:
remove hot files

heypoom committed:
add versioning

heypoom committed:
update rspack config

heypoom committed:
update rspack config

heypoom committed:
revert console.warn changes

heypoom committed:
add license error screen

heypoom committed:
cleanup createRoot

heypoom committed:
cleanup loading indicator

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx:
```diff
@@ -0,0 +1,86 @@
+
+  const hasEmbedTokenFeature = hasPremiumFeature("embedding_iframe_sdk");
+
+  // If the parent page is not running on localhost and
+  // the token feature is not present, we show an error message
+  if (!embedSettings._isLocalhost && !hasEmbedTokenFeature) {
+    return <SdkIframeInvalidLicenseError />;
+  }
```
We want to let folks without licenses play with the SDK in localhost. We check the origin of the embedder instead of the embed frame's origin, because most people will likely try out new iframe embedding with Cloud starter.

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/index.ts:
```diff
@@ -0,0 +1,6 @@
+import { PLUGIN_EMBEDDING_IFRAME_SDK } from "metabase/plugins";
+import { hasPremiumFeature } from "metabase-enterprise/settings";
+
+if (hasPremiumFeature("embedding_iframe_sdk")) {
+  PLUGIN_EMBEDDING_IFRAME_SDK.isEnabled = () => true;
+}
```
Will be used when the UI is introduced. For now it's mostly unused as we detect with `hasPremiumFeature` directly to avoid importing EE plugins into the isolated route.

heypoom commented on rspack.config.js:
```diff
@@ -377,4 +383,20 @@ if (devMode) {
+if (process.env.MB_EDITION === "ee") {
+  const SDK_IFRAME_EMBED_PLUGIN_PATH = `${ENTERPRISE_SRC_PATH}/embedding_iframe_sdk`;
+
+  if (config.entry) {
+    config.entry["app-embed-sdk"] =
+      `${SDK_IFRAME_EMBED_PLUGIN_PATH}/app-embed-sdk.tsx`;
+  }
```
Build an entry point for the `/embed/sdk/v1` route.

heypoom committed:
merge settings in embed tag

heypoom committed:
merge embed settings

WiNloSt commented on .gitignore:
```diff
@@ -59,8 +59,11 @@
 /resources/frontend_client/index.html
 /resources/frontend_client/public.html
+/resources/frontend_client/app/embed.v1.js
+/resources/frontend_client/app/main.*.hot*
```
just asking if it's possible to have users imported from `/app/dist/embed.v1.js` so we put everything in the same bucket?

That way, we don't have to have this exception. I imagine, if we put stuff directly on `/app` we might need to exclude more stuff in the future. And this doesn't look that maintainable.

What do you think?

Or at least, I'd expect `embed.v1.js` in `/app` but can it refer to other files in `/app/dist`?

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/bd189b7862dbba16903cc67643712c086603b74c - it now uses `/app/dist/embed.v1.js` and I reverted the gitignore

heypoom replied:
@npretto the iframeresizer is a static JS file (without build process from webpack and rspack) so that works directly; the problem is more with rspack's hot reloading sadly.

(I tried to disable hot reload but that's not possible as I was sharing the rspack process with the main bundle, and the first one takes precedence)

WiNloSt replied:
☝️ Good point. I'd expect users to just copy-paste all the external links though. I think the impact would be minimal, but it wouldn't hurt to stay consistent, if there's no repercussion.

heypoom replied:
I personally prefer `/app/embed.v1.js` too, I'll try to see if I can add a CopyPlugin and just copy the resulting file to `/app` or something similar

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/app-embed-sdk.tsx:
```diff
@@ -0,0 +1,27 @@
+
+  const rootElement = document.getElementById("root");
+
+  if (!rootElement) {
```
I wish there's a little bit of console log, or a comment for us just to know this shouldn't happen. i.e. we control how the HTML is served and this case shouldn't really happen. It's just that TypeScript being whiny and we need to satisfy it 😆

heypoom replied:
good point! let me add a console log

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/bd189b7862dbba16903cc67643712c086603b74c

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx:
```diff
@@ -0,0 +1,86 @@
+  }, [embedSettings]);
+
+  // The SDK will show its own loading indicator, so we don't need to show it twice.
+  if (embedSettings === null || !authConfig) {
```
it feels like we're mixing validation and the definition of `authConfig`. What about we check everything upfront, `embeddingSettings` couldn't be `null` or you could use `!embedSettings` like line 22. And check `!embedSettings.instanceUrl || !embedSettings.apiKey)` here too.

That way, we have a single location that returns `null`, well... you also have a single null return location, but `authConfig` also partially do validation, and I'd rather we have everything in one place, so I don't have to jump around when trying to understand this component.

heypoom replied:
That works! Let me move it up.

heypoom replied:
Thanks! Fixed in https://github.com/metabase/metabase/pull/56915/commits/0d6440e6b51125b2756143710fce7e6ddca10490. The reason that condition was needed was because of `authConfig` being `useMemo`'d, but I think we actually might _not_ need to memoize it from what I can see in `use-init-data.ts`.

WiNloSt replied:
this is much cleaner!

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/app-embed-sdk.tsx:
```diff
@@ -0,0 +1,27 @@
+import { StrictMode } from "react";
```
I'm not sure if the directory name is intended; we're mixing snake case and kebab case like crazy here. can we stick to just one convention if possible?

heypoom replied:
@WiNloSt It was intentional -- this follows the convention of the other `app-` entry points (e.g. `app-embed.js`). I'll move this file to live with its entry point friends to address your other concern of EE licensing 👍🏻

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/bd189b7862dbba16903cc67643712c086603b74c - now it lives in the correct directory and has the same convention (i.e. `app-embed.js`, `app-embed-sdk.tsx`)

WiNloSt commented on rspack.config.js:
```diff
@@ -377,4 +383,20 @@ if (devMode) {
+const additionalRspackConfig = [];
+
+if (process.env.MB_EDITION === "ee") {
+  const SDK_IFRAME_EMBED_PLUGIN_PATH = `${ENTERPRISE_SRC_PATH}/embedding_iframe_sdk`;
```
I'd consult some license experts here. Normally, we put stuff plugins under `metabase-enterprise` and we assign properties to plugins there. e.g. `PLUGIN_WHATEVER.property = xxx`

then we use this plugin in OSS code just fine.

For this case, we're technically importing EE code in an OSS file (it's not under enterprise directory), and we don't allow that on FE, so you can't import `EE` code inside OSS code (no `import xxx from 'metabase-enterprise/*` when we're inside `frontend/src/metabase/`.

Can we stick to the current approach? You can put `embedding_iframe_sdk` inside OSS file alongside `app-embed.js` but the components it renders are plugins.

heypoom replied:
That makes sense, let me move the entry point to live with its `app-*.js` friends then!

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/bd189b7862dbba16903cc67643712c086603b74c, moved to use EE plugins!

Alberto wants folks to be able to try out the iframe embedding on EE builds without a valid license, so I always assign the EmbedRoute component on the EE plugin. I then added a separate `hasValidLicense` field that is guarded behind `hasPremiumLicense(token)` that we can use.

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/plugins.js:
```diff
@@ -18,6 +18,7 @@ import "./database_routing";
 import "./whitelabel";
 import "./embedding";
 import "./embedding-sdk";
+import "./embedding_iframe_sdk";
```
i'm not convinced about the case. I think this should be kebab-case following other siblings here.

heypoom replied:
@WiNloSt actually `embedding-sdk` is wrong, all other EE plugin is `underscore_case` not `kebab-case`

![CleanShot 2568-05-09 at 23 05 07@2x](https://github.com/user-attachments/assets/2bbda834-7e08-41a2-80fc-471cd9f4e7fc)

WiNloSt replied:
You're absolutely right. Just keep it as is.

WiNloSt replied:
I created an issue for that https://linear.app/metabase/issue/EMB-383/make-ee-plugin-directory-name-cases-consistent

WiNloSt commented on frontend/src/metabase/redux/embed/embed.ts:
```diff
@@ -69,6 +69,9 @@ const interactiveEmbedSlice = createSlice({
         ...action.payload,
       };
     },
+    setIsEmbeddingSdk: (state, action: PayloadAction<boolean>) => {
```
I don't see we're using it here yet. If this is for future PR can we introduce it then?

heypoom replied:
@WiNloSt Good eye - this is actually a leftover code. I'll delete this!

WiNloSt reviewed (commented):
I'm 50/50 on the license. Can you consult someone about using the entrypoint from enterprise directory? I'll continue the rest of the PR next week 🙏

heypoom committed:
simplify licensing

heypoom committed:
fix enterprise source path

heypoom committed:
use ee plugins to resolve license

heypoom committed:
cleanup authConfig

heypoom committed:
fix build path

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+type AllowedEmbedSettingKey = (typeof ALLOWED_EMBED_SETTING_KEYS)[number];
+
+class MetabaseEmbed {
+  static readonly VERSION = "1.0.0";
```
When do we change this version? How does this version relate to any of the SDK versions? I guess they're totally different.

heypoom replied:
Yep!

- I think we update this whenever we make (potentially breaking) changes to the script tag, just for the sake of diagnostics. I think we can keep it simple by bumping the major version when we make a breaking change.
- Yes, it'll be totally different than the SDK version. The SDK component is essentially coupled to the Metabase instance's version they're using (e.g. the release-x.54.x branch when the Metabase binary is being built for that release), so it does not correlate with the version on NPM

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+  private iframe: HTMLIFrameElement | null = null;
+
+  constructor(settings: SdkIframeEmbedTagSettings) {
+    if (!settings.apiKey || !settings.instanceUrl) {
```
How does this check relate to https://github.com/metabase/metabase/blob/d908c8af7f07698cd5b367f6246d52b0bb6a8bac/enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx#L21

At glance, it feels like this check would never be true since we guard check before rendering this component.

heypoom replied:
- The `if (!settings.apiKey || !settings.instanceUrl) {` in the embed.js is for validation purposes.
- The `if (!embedSettings || !embedSettings.instanceUrl || !embedSettings.apiKey) {` in the iframe for data loading purposes -- `embedSettings` will be `null` for a long while until the `postMessage` communication completes (the iframe tells the script tag "we're ready", and the script tag sending the embed settings to the iframe).

heypoom replied:
Clarified as comments in https://github.com/metabase/metabase/pull/56915/commits/485a26e6adefdad649db69e94ecca1289965216f

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+  private iframe: HTMLIFrameElement | null = null;
+
+  constructor(settings: SdkIframeEmbedTagSettings) {
+    if (!settings.apiKey || !settings.instanceUrl) {
```
Why isn't this condition inside `_validateEmbedSettings`? It feels to me this belongs there.

heypoom replied:
This was because `_validateEmbedSettings` is also used for the `embed.updateSettings()` public method where the embedder can update their settings dynamically, and therefore `updateSettings({ theme })` would fail to validate because it does not have an `apiKey` and `instanceUrl`.

But now that I think about it, we can just do `_validateEmbedSettings` inside `_setEmbedSettings` after we merged the settings and the code will be much simpler, because `apiKey` and `instanceUrl` will always be there. Let me update this!

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/92ca062bfc9e07821972a4d22084a5bc9791d16f -- I moved this to `_validateEmbedSettings`

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+    this._settings = settings;
+    this._settings._isLocalhost = this.getIsLocalhost();
+
+    this._handleMessage = this._handleMessage.bind(this);
```
I'm 50/50 we don't need this. The reason being that 1. `_handleMessage` is defined using arrow function, so `this` is already bound. But I'm not sure if this will take effect immediately in the constructor since `_setup` also calls `_handleMessage`

Can you try removing this line to see if it's still working as expected?

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/b3edbee70938f4c5feb6f44be7dc4754f1b21448, thanks! I initially wrote it as a regular function and I must've moved it to arrow function later without cleaning this up.

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    if (!parentContainer) {
+      raiseError(`cannot find embed container "${target}"`);
```
should/could this be a part of `_validateEmbedSettings`? It just seems to me we spread condition checks everywhere.

heypoom replied:
Yep! I just realized we can just validate after we merge the settings, so we can bundle all the condition checks in a single method [1]. Let me update this.

[1] https://github.com/metabase/metabase/pull/56915/files#r2087879224

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+  /**
+   * Merge these settings with the current settings.
+   */
+  public updateSettings(settings: Partial<SdkIframeEmbedSettings>) {
```
where is this used?

heypoom replied:
@WiNloSt this is a public method that is used by the embedder in their website, which helps them to update the embed settings without reloading the whole iframe. I have an example in the end-to-end tests PR that covers this (also in the testing plan). For example, you can update the theme dynamically using this:

```tsx
embed.updateSettings({ theme: { colors: { brand: "#2d2d30" } }})
```

WiNloSt replied:
Ah right. Thanks for the clarification.

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/embed.v1.ts:
```diff
@@ -0,0 +1,170 @@
+    parentContainer.appendChild(this.iframe);
+  }
+
+  private getIsLocalhost() {
```
should/could this one be prefixed with `_` like other private method?

heypoom replied:
Absolutely! Lemme fix that

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/3099b7d158898d4b84c5efc444483847494e73ee, thanks!

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/hooks/use-sdk-iframe-embed-event-bus.ts:
```diff
@@ -0,0 +1,42 @@
+    return () => {
+      window.removeEventListener("message", messageHandler);
+    };
+  });
```
Is it intentional that we run this effect every render? I suppose we only do it once.

```suggestion
  }, []);
```

heypoom replied:
Good catch, can't believe I missed that.

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/56915/commits/3099b7d158898d4b84c5efc444483847494e73ee, thanks!

WiNloSt reviewed (approved):
I'll pre-approve, but please check with other comments if they're still valid and address as you see fit.

heypoom commented on enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk/components/SdkIframeEmbedRoute.tsx:
```diff
@@ -0,0 +1,74 @@
+  );
+};
+
+export const SdkIframeEmbedView = ({
```
Nice catch!

heypoom committed:
remove re-export from SdkIframeEmbedView

Co-authored-by: Mahatthana (Kelvin) Nomsawadi <me@bboykelvin.dev>

heypoom committed:
address reviews

heypoom committed:
clarify on embed settings

heypoom committed:
remove redundant api key check

heypoom committed:
fix ee plugins not loading

heypoom committed:
remove handleMessage binding

heypoom committed:
remove embed_sdk.html

heypoom committed:
fix merge conflicts

heypoom committed:
revert merge conflicted files

heypoom committed:
restore embedding-iframe-sdk feature flag

heypoom committed:
fix literal string lint error

heypoom committed:
fix crashes in the main app due to ee plugins

experiment -- comment

add EE overrides

experiment: disable Metabot

experiment: provide metabotPlugin with metabotInitialState

fix circular deps

use sdk-specific-imports instead

cleanup changes

rename sdk-overrides file to sdk-specific-imports

heypoom committed:
clarify why we import ee-plugins

heypoom committed:
restore incorrect merge conflict resolution

heypoom committed:
experiment: remove all realHover assertions

heypoom committed:
experiment: temporarily skip dark theme test
